### PR TITLE
feat(dbslice): Add self-laundering iterator in `DbSlice`

### PR DESCRIPTION
**Pull request by chakaz**
A self-laundering iterator will enable us to, eventually, yield from fibers while holding an iterator. For example:

```cpp
auto it1 = db_slice.Find(...);
Yield();  // Until now - this could have invalidated `it1`
auto it2 = db_slice.Find(...);
```

Why is this a good idea? Because it will enable yielding inside `PreUpdate()` which will allow breaking down of writing huge entries in small quantities to disk/network, eliminating the need to allocate huge chunks of memory just for serialization.

Also, it'll probably unlock future developments as well, as yielding can be useful in other contexts.

### CODE REVIEW

File: src/server/db_slice.h [hunk 1]
```diff
@@ -10,9 +10,35 @@
...
+
+  bool operator==(const StringOrView& o) const;
+  bool operator!=(const StringOrView& o) const;
+
+  std::string_view view() const;
+
+ private:
+  void Set(std::string s, std::string_view sv);
+
+  std::string s_;
```
Comment (romange) [inline src/server/db_slice.h:38]: nice. why not variant, btw?

File: src/server/db_slice.h [hunk 2]
```diff
@@ -69,6 +95,79 @@ class DbSlice {
...
+      return dfly::IsValid(it_);
+    }
+
+   private:
+    void LaunderIfNeeded() const {  // const is a lie
+      if (!dfly::IsValid(it_)) {
+        return;
+      }
+
+      assert(!key_.view().empty());
```
Comment (romange) [inline src/server/db_slice.h:152]: `set "" bar` will crash this

File: src/server/db_slice.h (hunk 1 above)
Comment (dranikpg) [reply src/server/db_slice.h:38]: > TODO: Move this, use it in all places we have variant<string, string_view>
 
Why do we need a big custom class if a simple wrapper is enough?  🤷🏻‍♂️ we have something similar here:

https://github.com/dragonflydb/dragonfly/blob/a93ad4e86f2e2a80e9d952bbd991ea954f0940fe/src/server/table.h#L102-L118

File: src/server/db_slice.h [hunk 3]
```diff
@@ -69,6 +95,79 @@ class DbSlice {
...
+
+    auto IsOccupied() const {
+      LaunderIfNeeded();
+      return it_.IsOccupied();
+    }
+
+    bool IsValid() const {
+      LaunderIfNeeded();
+      return dfly::IsValid(it_);
+    }
```
Comment (dranikpg) [inline src/server/db_slice.h:144]: Instead of accesing it_ you can call just GetInnerIt(), so you don't have to call Launder manually

File: src/server/db_slice.h (hunk 1 above)
Comment (chakaz) [reply src/server/db_slice.h]: lol, this is WIP, I didn't mean for anyone to review it yet :)

Re/ variant: I wanted to see an implementation with both members, where `s_` is only used as a storage medium, and `sv_` is always used for inspecting the string. But I agree that `variant` is better (except for an added `if` in the `visit` path, but that's super unimportant and premature optimization).

Re/ big custom class - I'm not sure what you mean by big, but the `Key` you linked to is also a "custom class" :rofl: It's just a private class that's not usable outside of table, and is one of this things that should be replaced by a general utility.

Anyway, I moved the `StringOrView` class to a standalone file, took the `variant` approach like `Key` and moved table.h to use it.

But, as I said, this is still WIP, so unless you are really bored, I wouldn't look at this PR yet :nerd_face:

File: src/server/db_slice.h (hunk 2 above)
Comment (chakaz) [reply src/server/db_slice.h:128]: good catch :)
(but it doesn't happen with the new `variant` impl)

File: src/server/db_slice.h (hunk 3 above)
Comment (chakaz) [reply src/server/db_slice.h:120]: Good idea!

**Review by dranikpg — DISMISSED** [review]
I can't guarantee that every replacement was correct 😅 But the ones I looked at were ok

File: src/server/snapshot.cc [hunk 4]
```diff
@@ -242,7 +242,8 @@ bool SliceSnapshot::BucketSaveCb(PrimeIterator it) {
     ++stats_.skipped;
     return false;
   }
-  db_slice_->FlushChangeToEarlierCallbacks(current_db_, it, snapshot_version_);
+  db_slice_->FlushChangeToEarlierCallbacks(current_db_, DbSlice::Iterator::FromPrime(it),
```
Comment (adiholden) [inline src/server/snapshot.cc:245]: why do we need to call DbSlice::Iterator::FromPrime here?

File: src/server/db_slice.cc [hunk 5]
```diff
@@ -334,29 +337,9 @@ void DbSlice::AutoUpdater::Run() {
     return;
   }
 
-  // Check that AutoUpdater does not run after a key was removed.
-  // If this CHECK() failed for you, it probably means that you deleted a key while having an auto
-  // updater in scope. You'll probably want to call Run() (or Cancel() - but be careful).
-  DCHECK(IsValid(fields_.db_slice->db_arr_[fields_.db_ind]->prime.Find(fields_.key)))
```
Comment (adiholden) [inline src/server/db_slice.cc:340]: why did you remove this check?
can we run now autoupdater after the key was removed?

File: src/core/string_or_view.cc [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  return sov;
+}
+
+StringOrView StringOrView::FromView(string_view sv) {
+  StringOrView sov;
+  sov.val_ = sv;
+  return sov;
+}
+
+bool StringOrView::operator==(const StringOrView& o) const {
```
Comment (romange) [inline src/core/string_or_view.cc:22]: nit: i would move onelines to class definition and others to the header file and get rid of cc.

File: src/server/generic_family.cc [hunk 7]
```diff
@@ -288,8 +288,8 @@ void Renamer::Find(Transaction* t) {
     auto& db_slice = EngineShard::tlocal()->db_slice();
     auto [it, exp_it] = db_slice.FindReadOnly(t->GetDbContext(), res->key);
 
-    res->found = IsValid(it);
```
Comment (romange) [inline src/server/generic_family.cc:291]: nit: you could provide `IsValid` override in db_slice.h and greatly reduce the size of the PR.

Comment (romange) [reply src/server/generic_family.cc:291]: btw, this is why it existed in the first place - I had different type of iterators in the past and wanted to handle them "cleanly".

File: src/server/db_slice.h [hunk 8]
```diff
@@ -69,6 +71,77 @@ class DbSlice {
...
+    auto IsOccupied() const {
+      return GetInnerIt().IsOccupied();
+    }
+
+    bool IsValid() const {
+      return dfly::IsValid(GetInnerIt());
+    }
+
+   private:
+    void LaunderIfNeeded() const {  // const is a lie
```
Comment (romange) [inline src/server/db_slice.h:121]: nit: to pull the definition below DbSlice declaration.

File: src/server/detail/table.h [hunk 9]
```diff
@@ -53,6 +53,10 @@ struct ExpireTablePolicy {
     return s.HashCode();
   }
 
+  static uint64_t HashFn(std::string_view u) {
```
Comment (romange) [inline src/server/detail/table.h:56]: why this was needed?

File: src/core/string_or_view.h [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  std::string_view view() const;
+
+  friend std::ostream& operator<<(std::ostream& o, const StringOrView& key) {
+    return o << key.view();
+  }
+
+  // Make hashable
+  template <typename H> friend H AbslHashValue(H h, const StringOrView& c) {
+    return H::combine(std::move(h), c.view());
+  }
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/core/string_or_view.h:38]: Roman suggest removing the cc file as well, no?

[the remark this answers is not in the document]
Comment (chakaz) [reply src/core/string_or_view.h:38]: Yes, both you and Roman think it should be removed so I'll remove it. That's what I tried to say :upside_down_face:

File: src/server/table.cc [hunk 11]
```diff
@@ -57,37 +57,32 @@ SlotStats& SlotStats::operator+=(const SlotStats& o) {
...
 
 bool LockTable::Acquire(std::string_view key, IntentLock::Mode mode) {
   DCHECK_EQ(KeyLockArgs::GetLockKey(key), key);
 
-  auto [it, inserted] = locks_.try_emplace(Key{key});
-  if (!inserted)            // If more than one transaction refers to a key
-    it->first.MakeOwned();  // we must fall back to using a self-contained string
+  auto [it, inserted] = locks_.try_emplace(Key::FromView(key));
+  if (!inserted)                              // If more than one transaction refers to a key
+    const_cast<Key&>(it->first).MakeOwned();  // we must fall back to using a self-contained string
```
[the remark this answers is not in the document]
Comment (chakaz) [reply src/server/table.cc:77]: fwiw it's exactly the same as before, except that now `MakeOwned()` is non-`const` (because it's not a `const` operation), and we do the `const_cast` because know that it wouldn't matter

File: src/server/snapshot.cc (hunk 4 above)
Comment (chakaz) [reply src/server/snapshot.cc:245]: `FlushChangeToEarlierCallbacks` calls pre-update callbacks. I _think_ (but I'm not sure) that we will need to give them the auto-laundering iterator as well, as they might also yield.
wdyt?

File: src/server/db_slice.cc (hunk 5 above)
Comment (chakaz) [reply src/server/db_slice.cc:340]: Nope, I should not have removed it.

File: src/server/generic_family.cc (hunk 7 above)
Comment (chakaz) [reply src/server/generic_family.cc:291]: oh, i forgot about this comment. Please do not review yet, I'll fix this as well.

File: src/server/snapshot.cc (hunk 4 above)
Comment (romange) [reply src/server/snapshot.cc:245]: is not the whole motivation of this PR is to allow `SliceSnapshot::OnDbChange` to yield?

Comment (chakaz) [reply src/server/snapshot.cc:245]: We definitely want OnDbChange to be able to yield. And I think it's a good idea that it will have an auto laundering iterator to make things easier there. But it's not a necessity as the callbacks could launder automatically and anyway they should be using the values and not find etc.
but again, I think it's a good idea.

**Comment by chakaz** [comment]
@romange / @adiholden friendly ping :)

File: src/server/db_slice.h [hunk 12]
```diff
@@ -330,12 +391,13 @@ class DbSlice {
   }
 
   // Check whether 'it' has not expired. Returns it if it's still valid. Otherwise, erases it
-  // from both tables and return PrimeIterator{}.
+  // from both tables and return Iterator{}.
   struct ItAndExp {
     PrimeIterator it;
```
Comment (adiholden) [inline src/server/db_slice.h:396]: Shouldnt we change this as well PrimeIterator ->  Iterator and ExpireIterator -> ExpIterator ?

Comment (chakaz) [reply src/server/db_slice.h:396]: The thing is, we call this also in places where we have just `PrimeIterator` in our hands, like when we search directly in the table (not through DbSlice). I thought it would be better to leave as is, as it won't have any benefit to change.
wdyt?

File: src/server/generic_family.cc [hunk 13]
```diff
@@ -1500,7 +1500,7 @@ OpStatus GenericFamily::OpMove(const OpArgs& op_args, string_view key, DbIndex t
   db_slice.ActivateDb(target_db);
 
   bool sticky = from_res.it->first.IsSticky();
-  uint64_t exp_ts = db_slice.ExpireTime(from_res.exp_it);
+  uint64_t exp_ts = db_slice.ExpireTime(from_res.exp_it.GetInnerIt());
```
Comment (adiholden) [inline src/server/generic_family.cc:1503]: I think that you can call db_slice.ExpireTime(from_res.exp_it); right?

File: src/server/string_family.cc [hunk 14]
```diff
@@ -531,7 +529,7 @@ SinkReplyBuilder::MGetResponse OpMGet(bool fetch_mcflag, bool fetch_mcver, const
       }
 
       if (fetch_mcver) {
-        resp.mc_ver = it.GetVersion();
+        resp.mc_ver = it.GetInnerIt().GetVersion();
```
Comment (adiholden) [inline src/server/string_family.cc:532]: nit: it->GetVersion()

File: src/server/db_slice.h [hunk 15]
```diff
@@ -518,4 +580,36 @@ class DbSlice {
...
+template <typename T> void DbSlice::IteratorT<T>::LaunderIfNeeded() const {
+  if (!dfly::IsValid(it_)) {
+    return;
+  }
+
+  // We currently do not shrink dashtables, but when we will, we must also check if "it.segment_id"
+  // falls into a valid range.
+  uint64_t current_epoch = util::fb2::FiberSwitchEpoch();
+  if (current_epoch != fiber_epoch_) {
+    if (it_->first != key_.view()) {
```
Comment (adiholden) [inline src/server/db_slice.h:608]: !it->IsOccupied() || it_->first != key_.view()

File: src/server/db_slice.h (hunk 12 above)
Comment (adiholden) [reply src/server/db_slice.h:396]: Well its confusing and can lead to bugs but I think we can revisit this later

Comment (chakaz) [reply src/server/db_slice.h:396]: Can you give an example for a bug it could lead to?
Because it's strongly typed, I'm not sure what you think of. I am happy to change it now instead of later though.

Comment (adiholden) [reply src/server/db_slice.h:400]: If we have places in which we call ExpireIfNeeded when we call with Iterator and not PrimeIterator and than we will do another Find call. I believe we dont have such cases right now but the fact that we do such api that we enable calling ExpireIfNeeded with Iterator can lead to such bugs.

Comment (chakaz) [reply src/server/db_slice.h]: Right, that makes sense. Done.

File: src/server/db_slice.h [hunk 16]
```diff
@@ -605,7 +609,7 @@ template <typename T> void DbSlice::IteratorT<T>::LaunderIfNeeded() const {
   // falls into a valid range.
   uint64_t current_epoch = util::fb2::FiberSwitchEpoch();
   if (current_epoch != fiber_epoch_) {
-    if (it_->first != key_.view()) {
+    if (!it_.IsOccupied() || it_->first != key_.view()) {
```
Comment (adiholden) [inline src/server/db_slice.h:616]: now you can remove the above comment as IsOccupied will check if we still fall into valid range

**Review by adiholden — APPROVED** [review]

**Review by adiholden — APPROVED** [review]
